src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -378,6 +383,14 @@ protected override bool Read(ReadableBuffer readableBuffer, WritableBuffer writa
 
                 return _inputLength == 0;
             }
+
+            protected override void OnReadStart()
+            {
+                if (_contentLength > _context.MaxRequestBodySize)
+                {
+                    _context.RejectRequest(RequestRejectionReason.RequestBodyTooLarge);
```

Comment:
I've been misled by this too. At one point, someone compared this to explicit throws (e.g. `throw MethodThatCreatesException()`) and concluded that it doesn't inline as well, enough that it hurts perf. I think that was @pakrym or @benaadams